pkg/virt-controller/services/renderresources.go

```diff
@@ -127,35 +119,37 @@ func addToCPU(resource map[k8sv1.ResourceName]resource.Quantity, q resource.Quan
 	}
 }
 
-func WithoutDedicatedCPU(cpu *v1.CPU, cpuAllocationRatio int, withCPULimits bool) ResourceRendererOption {
+// Helper function to extract IO thread CPU count from VMI
+func getIOThreadsCount(vmi *v1.VirtualMachineInstance) int64 {
+	if vmi == nil || vmi.Spec.Domain.IOThreads == nil ||
+		vmi.Spec.Domain.IOThreads.SupplementalPoolThreadCount == nil ||
+		*vmi.Spec.Domain.IOThreads.SupplementalPoolThreadCount < 1 {
+		return 0
+	}
+	return int64(*vmi.Spec.Domain.IOThreads.SupplementalPoolThreadCount)
+}
+
+func WithoutDedicatedCPU(vmi *v1.VirtualMachineInstance, cpuAllocationRatio int, withCPULimits bool) ResourceRendererOption {
 	return func(renderer *ResourceRenderer) {
+		cpu := vmi.Spec.Domain.CPU
 		vcpus := calcVCPUs(cpu)
-		if vcpus != 0 && cpuAllocationRatio > 0 {
-			val := float64(vcpus) / float64(cpuAllocationRatio)
+		ioThreadCPUs := getIOThreadsCount(vmi) // Get IO thread count
+		totalCPUs := vcpus + ioThreadCPUs      // Include IO threads
```

Comment:
So we'll dedicate another CPU per IO thread?

Doesn't this have the potential to dramatically increase resource consumption by existing VMs?
Are the implications of defining IO threads clear to the admin?

Reply:
Yes, that's how it's supposed to work. Similar to the emulator thread, we need to overallocate a physical CPU for each IOThread. 
I assume that was explained with the implementation of the original feature. This is a bug fix that corrects the malfunction.

Reply:
> Similar to the emulator thread, we need to overallocate a physical CPU for each IOThread.

Yeah, we allocate another pCPU for the emulator threads just because we (currently) don't have a better way of isolating it. I thought we'll use the same single extra core for ALL iothreads.

I'm just a bit concerned about existing VMs, and about this potentially not being clear to the VM owner.

Reply:
Right, but in this case, the IOThreads are as important (for being isolated) as vCPU and not as the emulator thread (which we isolate from interfering with the vCPUs)
Here is the original PR: https://github.com/kubevirt/kubevirt/pull/13110 and the blog that speaks about the importance of such isolation https://developers.redhat.com/articles/2024/09/05/scaling-virtio-blk-disk-io-iothread-virtqueue-mapping#how_iothread_virtqueue_mapping_works
`IOThreads need to respond quickly when the guest submits I/O. Therefore they should not compete for CPU time with the guest’s vCPU threads.`

Reply:
Our user-guide says [[1](https://kubevirt.io/user-guide/storage/disks_and_volumes/#iothreads-with-qemu-emulator-thread-and-dedicated-pinned-cpus)]:
> To further improve the vCPUs latency, KubeVirt can allocate an additional dedicated physical CPU[1](https://kubevirt.io/user-guide/compute/virtual_hardware/#cpu), exclusively for the emulator thread, to which it will be pinned. This will effectively "isolate" the emulator thread from the vCPUs of the VMI. When ioThreadsPolicy is set to auto IOThreads will also be "isolated" from the vCPUs and placed on the same physical CPU as the QEMU emulator thread.

And `supplementalPool` is not mentioned.
Let's make sure we properly document this to avoid confusion.

Reply:
@alicefr, did you have a documentation PR by chance?

Reply:
@vladikr no I still need to add it, I have completely forgot it

Reply:
Thanks for the clarification @alicefr! :) It just got me a bit confused for a second.

> I will document this, I apologize to have for not having done immediately.

Awesome, no worries!
Feel free to cc me on the docs PR :+1: